nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-reporting-task/src/main/java/org/apache/nifi/reporting/riemann/RiemannReportingTask.java

```diff
@@ -0,0 +1,244 @@
+            transport = Transport.valueOf(context.getProperty(TRANSPORT_PROTOCOL).getValue());
+            String host = context.getProperty(RIEMANN_HOST).evaluateAttributeExpressions().getValue();
+            int port = context.getProperty(RIEMANN_PORT).asInteger();
+            RiemannClient client = null;
+            try {
+                switch (transport) {
+                case TCP:
+                    client = RiemannClient.tcp(host, port);
+                    break;
+                case UDP:
+                    client = RiemannClient.udp(host, port);
+                    break;
+                }
+                client.connect();
+                riemannClient = client;
+            } catch (IOException e) {
+                if (client != null) {
+                    client.close();
+                }
+                throw new ProcessException(String.format("Unable to connect to Riemann [%s:%d] (%s)\n%s", host, port, transport, e.getMessage()));
+            }
+        }
+        metricsService = new MetricsService(context.getProperty(SERVICE_PREFIX).getValue(), context.getProperty(HOSTNAME).evaluateAttributeExpressions().getValue(),
+                context.getProperty(TAGS).getValue().split(","));
+
+        // Add dynamic properties as attributes to messages
+        Map<String, String> additionalProperties = Maps.newHashMapWithExpectedSize(context.getProperties().size());
+        for (PropertyDescriptor propertyDescriptor : context.getProperties().keySet()) {
+            if (propertyDescriptor.isDynamic()) {
+                additionalProperties.put(propertyDescriptor.getName(), context.getProperty(propertyDescriptor).getValue());
+            }
+        }
+        metricsService.setAdditionalAttributes(additionalProperties);
+    }
+
+    /**
+     * Retrieves all new bulletins in the `BulletinRepository` since the last time this method was called.
+     *
+     * @param context
+     *          Reporting context
+     * @return new bulletins
+     */
+    private List<Bulletin> getNewBulletins(ReportingContext context) {
+        BulletinQuery.Builder bulletinQueryBuilder = new BulletinQuery.Builder().messageMatches(".*").after(lastObservedBulletinId);
+        List<Bulletin> bulletins = context.getBulletinRepository().findBulletins(bulletinQueryBuilder.build());
+        List<Bulletin> filteredBulletins = Lists.newArrayListWithCapacity(bulletins.size());
+        int logLevelValue = LogLevel.valueOf(context.getProperty(MIN_BULLETIN_LEVEL).getValue()).getValue();
+        if (bulletins.size() > 0) {
+            for (Bulletin bulletin : bulletins) {
+                if (LogLevel.valueOf(bulletin.getLevel()).getValue() >= logLevelValue) {
+                    filteredBulletins.add(bulletin);
+                }
+            }
+            for (Bulletin bulletin : bulletins) {
+                lastObservedBulletinId = Math.max(lastObservedBulletinId, bulletin.getId());
```

Comment:
Seems like this line could be incorporated into a loop above instead of looping twice, right?